File: LeetCode/416.py
# Brute Force / Naive
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        self.res = False
        self.t = sum(nums)

        if self.t % 2 != 0:
            return False

        def dfs(i, cand):
            if cand == self.t//2:
                self.res = True
                return
            if i >= len(nums):
                return
            dfs(i+1, cand+nums[i])
            dfs(i+1, cand)
        
        dfs(0, 0)
        return self.res

# DP
class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        t = sum(nums)
        if t % 2:
            return False

        dp = set()

        for i in range(len(nums)):
            for n in dp.copy():
                dp.add(nums[i]+n)
            dp.add(nums[i])

        return True if t // 2 in dp else False

File: leetcode/416.cpp
// Brute Force
class Solution
{
private:
    int target = 0;
    vector<int> nums;

public:
    bool canPartition(vector<int> &nums)
    {
        this->nums = nums;
        for (int num : nums)
        {
            target += num;
        }
        if (target % 2)
            return false;
        target /= 2;
        return dfs(0, 0);
    }

    bool dfs(int idx, int sum)
    {
        if (sum == target)
            return true;
        if (idx >= nums.size())
            return false;
        return dfs(idx + 1, sum + nums[idx]) || dfs(idx + 1, sum);
    }
};

// Memoisation
class Solution
{
private:
    int target = 0;
    int n;
    vector<vector<int>> dp;
    vector<int> nums;

public:
    bool canPartition(vector<int> &nums)
    {
        this->nums = nums;
        this->n = nums.size();

        for (int num : nums)
        {
            target += num;
        }
        if (target % 2)
            return false;
        target /= 2;
        this->dp = vector<vector<int>>(n, vector<int>(target + 1, -1));
        return dfs(0, 0);
    }

    bool dfs(int idx, int sum)
    {
        if (sum == target)
            return true;
        if (idx >= n)
            return false;
        if (dp[idx][sum] != -1)
            return dp[idx][sum];

        bool take = false;
        bool ntake = dfs(idx + 1, sum);
        if (nums[idx] + sum <= target)
            take = dfs(idx + 1, sum + nums[idx]);
        return dp[idx][sum] = take || ntake;
    }
};
